src/briefcase/platforms/macOS/__init__.py

```diff
@@ -577,48 +583,67 @@ def package_app(
         """Package an app bundle.
 
         :param app: The application to package
-        :param sign_app: Should the application be signed? Default: ``True``
         :param notarize_app: Should the app be notarized? Default: ``True`` if the
             app has been signed with a real identity; ``False`` if the app is
             unsigned, or an ad-hoc signing identity has been used.
         :param identity: The code signing identity to use. This can be either
             the 40-digit hex checksum, or the string name of the identity.
             If unspecified, the user will be prompted for a code signing
-            identity. Ignored if ``sign_app`` is ``False``.
-        :param adhoc_sign: If ``True``, code will be signed with adhoc identity of "-"
+            identity. Ignored if ``adhoc_sign`` is ``True``.
+        :param adhoc_sign: If ``True``, code will be signed with ad-hoc identity
+            of "-", and the resulting app will not be re-distributable.
         """
-        if sign_app:
-            if adhoc_sign:
+        if adhoc_sign:
+            if notarize_app:
+                raise BriefcaseCommandError(
+                    "Can't notarize an app with an ad-hoc signing identity"
+                )
+
+            identity = "-"
+            self.logger.info(
+                "Signing app with ad-hoc identity...",
+                prefix=app.app_name,
+            )
+            self.logger.warning(
+                (
+                    "Because you are signing with the ad-hoc identity, this "
+                    "app will run, but cannot be re-distributed."
+                ),
+                prefix=app.app_name,
+            )
+        else:
+            identity, identity_name = self.select_identity(identity=identity)
+
+            if identity == "-":
                 if notarize_app:
                     raise BriefcaseCommandError(
-                        "Can't notarize an app with an adhoc signing identity"
+                        "Can't notarize an app with an ad-hoc signing identity"
                     )
-
-                identity = "-"
                 self.logger.info(
-                    "Signing app with adhoc identity...", prefix=app.app_name
+                    "Signing app with ad-hoc identity...",
+                    prefix=app.app_name,
+                )
+                self.logger.warning(
+                    (
+                        "Because you are signing with the ad-hoc identity, this "
+                        "app will run, but cannot be re-distributed."
+                    ),
+                    prefix=app.app_name,
```

Comment:
The duplication of the adhoc warnings here suggests there's an opportunity to clean up the logic - could we restrict the initial logic to `if adhoc: identity = '-'; else identity= select_identity()`, and then do all the validation assuming that we *have* a signing identity?